ast/format_test.go

```diff
@@ -0,0 +1,74 @@
+package ast_test
+
+import (
+	"github.com/google/go-cmp/cmp"
+	"github.com/influxdata/flux/semantic"
+	"github.com/influxdata/flux/semantic/semantictest"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/influxdata/flux/ast"
+	"github.com/influxdata/flux/parser"
+)
+
+func withEachFluxFile(t *testing.T, fn func(caseName, fileContent string)) {
+	dir, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+	path := filepath.Join(dir, "testdata")
+
+	fluxFiles, err := filepath.Glob(filepath.Join(path, "*.flux"))
+	if err != nil {
+		t.Fatalf("error searching for Flux files: %s", err)
+	}
+
+	for _, fluxFile := range fluxFiles {
+		ext := filepath.Ext(fluxFile)
+		prefix := fluxFile[0 : len(fluxFile)-len(ext)]
+		_, caseName := filepath.Split(prefix)
+
+		content, err := ioutil.ReadFile(fluxFile)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		fn(caseName, string(content))
+	}
+}
+
+func TestFormat(t *testing.T) {
+	// we compare the semantic (we also check that we got a valid query)
+	withEachFluxFile(t, func(caseName, content string) {
+		t.Run(caseName, func(t *testing.T) {
+			originalProgram, err := parser.NewAST(content)
+			if err != nil {
+				t.Fatal(errors.Wrapf(err, "original program has bad syntax:\n%s", content))
+			}
+
+			want, err := semantic.New(originalProgram)
+			if err != nil {
+				t.Fatal(errors.Wrapf(err, "original program is not a valid flux query: %s", content))
+			}
+
+			stringResult := ast.Format(originalProgram)
+
+			newProgram, err := parser.NewAST(stringResult)
+			if err != nil {
+				t.Fatal(errors.Wrapf(err, "new program has bad syntax:\n%s", stringResult))
+			}
+
+			got, err := semantic.New(newProgram)
+			if err != nil {
+				t.Fatal(errors.Wrapf(err, "original program is not a valid flux query: %s", stringResult))
+			}
+
+			if !cmp.Equal(want, got, semantictest.CmpOptions...) {
+				t.Errorf("to string conversion error:\nin:\t\t%s\nout:\t\t%s\n", content, stringResult)
+			}
```

Comment:
Can you remove the testing code around the semantic package? We want these tests for be focused on testing ast.Format only. This test should compare that the input file is the same as the formatted output.

Reply:
I used semantic for two reasons:

 - verify that the query is valid;
 - two statements can be expressed in different ways, but semantically, they mean the same thing. E.g.: `filter(fn: (r) => body)` is the same as `filter(fn: (r) => (body))`, or the float `1.0` is the same as `1.0000`, while they are different strings.

Should I input queries such that I can reproduce an identical output and compare strings?

Reply:
The above statements are true, but this test doesn't need to test those things, they are tested elsewhere, for example in the semantic package. Specifically the AST can represent an invalid query, but it should still be able to be formatted correctly.

For example:
```
filter(fn: false)
```

That is valid syntax and should be able to be formatted correctly, but it is not a valid query. But this test is only concerned with formatting not whether the script itself makes any sense.

Therefore this test should be testing that the output is as expected, which means we need to test the actual string output of Format. As for your question:


> Should I input queries such that I can reproduce an identical output and compare strings?

There are two ways to do this:

1. Input correctly formatted scripts and ensure that the output is identical
2. Have each test case have an input and output script. The input script can be formatted in the non "standard" way while the output will always be formatted according to the "standard"

I prefer option 2. It is easier to do this if you do not use files as the source data, rather just a list of structs that have two strings. For example you could follow this pattern here https://github.com/influxdata/flux/blob/master/semantic/types_test.go#L15